activestorage/lib/active_storage/service/s3_service.rb

```diff
@@ -137,8 +137,11 @@ def stream(key)
 
         raise ActiveStorage::FileNotFoundError unless object.exists?
 
-        while offset < object.content_length
-          yield object.get(range: "bytes=#{offset}-#{offset + chunk_size - 1}").body.string.force_encoding(Encoding::BINARY)
+        content_length = object.content_length
+        while offset < content_length
+          # Upper bound should not exceed content_length-1
+          upper_bound = [offset + chunk_size - 1, content_length - 1].min
```

Comment:
```suggestion
          chunk_size = [chunk_size, content_length - offset].min
```

Reply:
If so, then should we revert next line to origin?

Reply:
Yes, please.

Reply:
I believe explicit calculation of `upper_bound` will be more clear...